core/drivers/crypto/hisilicon/hpre_dh.c

```diff
@@ -0,0 +1,375 @@
+		return HISI_QM_DRVCRYPT_ENOMEM;
+	}
+
+	msg->x_p = data;
+	msg->x_p_dma = virt_to_phys(msg->x_p);
+
+	msg->g = msg->x_p + (msg->key_bytes << 1);
+	msg->g_dma = msg->x_p_dma + (msg->key_bytes << 1);
+	msg->out = msg->g + msg->key_bytes;
+	msg->out_dma = msg->g_dma + msg->key_bytes;
+
+	return 0;
+}
+
+static void hpre_dh_params_free(struct hpre_dh_msg *msg)
+{
+	if (msg->x_p) {
+		memzero_explicit(msg->x_p, msg->key_bytes << 1);
+		free(msg->x_p);
+		msg->x_p = NULL;
+	}
+}
+
+static int32_t hpre_dh_params_bn2bin(struct hpre_dh_msg *msg,
+				     struct dh_keypair *key,
+				     struct bignum *pubkey)
+{
+	uint8_t *p = msg->x_p + msg->key_bytes;
+	uint8_t *x = msg->x_p;
+	int32_t ret = 0;
+
+	msg->xbytes = BITS_TO_BYTES(key->xbits);
+	msg->out_bytes = msg->pbytes;
+	crypto_bignum_bn2bin(key->x, x);
+	crypto_bignum_bn2bin(key->p, p);
+
+	if (!pubkey) {
+		msg->gbytes = crypto_bignum_num_bytes(key->g);
+		crypto_bignum_bn2bin(key->g, msg->g);
+	} else {
+		msg->gbytes = crypto_bignum_num_bytes(pubkey);
+		if (msg->gbytes != msg->pbytes)
+			return HISI_QM_DRVCRYPT_EINVAL;
+		crypto_bignum_bn2bin(pubkey, msg->g);
+	}
+
+	ret = hpre_bin_from_crypto_bin(x, x, msg->key_bytes, msg->xbytes);
+	if (ret) {
+		EMSG("Fail to transfer dh_x from crypto_bin to hpre_bin");
+		return ret;
+	}
+
+	ret = hpre_bin_from_crypto_bin(p, p, msg->key_bytes, msg->pbytes);
+	if (ret) {
+		EMSG("Fail to transfer dh_p from crypto_bin to hpre_bin");
+		return ret;
+	}
+
+	ret = hpre_bin_from_crypto_bin(msg->g, msg->g, msg->key_bytes,
+				       msg->gbytes);
+	if (ret)
+		EMSG("Fail to transfer dh_g from crypto_bin to hpre_bin");
+
+	return ret;
+}
+
+static TEE_Result hpre_dh_request_init(struct hpre_dh_msg *msg,
+				       struct dh_keypair *key,
+				       struct bignum *pubkey)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	msg->alg_type = HPRE_ALG_DH;
+	msg->key_bytes = round_key_size_to_hw_size(msg->pbytes);
+	if (!msg->key_bytes)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	ret = hpre_dh_params_alloc(msg);
+	if (ret)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	ret = hpre_dh_params_bn2bin(msg, key, pubkey);
+	if (ret) {
+		hpre_dh_params_free(msg);
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static void hpre_dh_request_deinit(struct hpre_dh_msg *msg)
+{
+	hpre_dh_params_free(msg);
+}
+
+static TEE_Result hpre_dh_gen_keypair(struct dh_keypair *key,
+				      struct bignum *q __unused,
+				      size_t key_size)
+{
+	struct hpre_dh_msg msg = { };
+	TEE_Result ret = TEE_SUCCESS;
+	size_t p_bits = 0;
+
+	if (!key || !key->g || !key->p || !key->x || !key->y) {
+		EMSG("Invalid dh_gen_keypair input parameters");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	p_bits = crypto_bignum_num_bits(key->p);
+	if (!p_bits) {
+		EMSG("p_bits can not be zero");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+	msg.pbytes = BITS_TO_BYTES(p_bits);
+
+	if (!key_size) {
+		/* xbits */
+		key->xbits = p_bits;
+		ret = hpre_dh_gen_privkey(key->x, key->xbits);
```

Comment:
`hpre_dh_gen_privkey()` does not return a `TEE_Result` value.

Same comment for lines 300, 306, 339 and 345.

Reply:
![image](https://github.com/OP-TEE/optee_os/assets/118411745/a42a3886-481e-4fa3-8ee7-a38edd5cf3f0)
why does hpre_dh_gen_privkey()  not return a TEE_Result value?

Reply:
crypto_bignum_bin2bn returns a TEE_Result value,and TEE_ERROR_NO_DATA is also a TEE_Result value